src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ThemeManager.cs

```diff
@@ -1,278 +1,477 @@
+    private static void ApplyStyleOnWindow(Window window, bool useLightColors)
     {
-        ArgumentNullException.ThrowIfNull(dictionaryUri, nameof(dictionaryUri));
-
-        var newDictionary = new ResourceDictionary() { Source = dictionaryUri };
+        if(window == null || window.IsDisposed) return;
 
-        FindFluentThemeResourceDictionary(out ResourceDictionary fluentDictionary);
-        
-        if (fluentDictionary != null)
+        // We only apply Style on window, if the Window.Style has not already been set to avoid overriding users setting. 
+        if(window.Style == null)
         {
-            Application.Current.Resources.MergedDictionaries.Remove(fluentDictionary);
+            window.SetResourceReference(FrameworkElement.StyleProperty, typeof(Window));
         }
 
-        Application.Current.Resources.MergedDictionaries.Add(newDictionary);
+        window.SetImmersiveDarkMode(!useLightColors);
+
+        if(SystemParameters.HighContrast)
+        {
+            WindowBackdropManager.SetBackdrop(window, WindowBackdropType.None);
+        }
+        else
+        {
+            WindowBackdropManager.SetBackdrop(window, WindowBackdropType.MainWindow);
+        }
     }
 
     #endregion
 
-    #endregion
 
     #region Internal Properties
 
-    internal static bool IsFluentThemeEnabled => _isFluentThemeEnabled;
-    // TODO : Find a better way to deal with different default font sizes for different themes.
+    internal static bool DeferSyncingThemeModeAndResources { get; set; } = true;
+
+    internal static bool IsFluentThemeEnabled
+    {
+        get
+        {
+            if(Application.Current == null) return false;
+            return Application.Current.ThemeMode != ThemeMode.None;
+        }
+    }
+
+    internal static bool DeferredAppThemeLoading { get; set; } = false;
+
+    internal static bool IgnoreAppResourcesChange { get; set; } = false;
+
     internal static double DefaultFluentThemeFontSize => 14;
 
+    internal static WindowCollection FluentEnabledWindows { get; set; } = new WindowCollection();
+
     #endregion
 
-    #region Private Methods
 
-    private static Uri GetFluentThemeResourceUri(string systemTheme, bool useLightMode)
+    #region Helper Methods
+
+    private static bool GetUseLightColors(ThemeMode themeMode)
+    {
+        // Is this needed ?
+        if(themeMode == ThemeMode.None)
+        {
+            return true;
+        }
+
+        // Do we need to add a check for ThemeMode.None theme?
+        return themeMode == ThemeMode.Light || (themeMode == ThemeMode.System && IsSystemThemeLight());
+    }
+
+    private static ThemeMode GetThemeModeFromResourceDictionary(ResourceDictionary rd)
     {
-        string themeFileName = "Fluent." + (useLightMode ? "Light" : "Dark") + ".xaml";
+        ThemeMode themeMode = ThemeMode.None;
+
+        if (rd == null) return themeMode;
+
+        int index = FindLastFluentThemeResourceDictionaryIndex(rd);
```

Comment:
same here TryGet... would work as well

Reply:
Actually, we are using the index below to extract the source uri of the merged dictionary. Leaving it as it is.